platform/src/main/java/org/stellar/anchor/platform/callback/RestRateIntegration.java

```diff
@@ -68,29 +88,186 @@ public GetRateResponse getRate(GetRateRequest request) throws AnchorException {
-      throw new ServerErrorException("internal server error");
+      logErrorAndThrow("missing 'price' in the GET /rate response", ServerErrorException.class);
     }
 
-    if (request.getType() != GetRateRequest.Type.INDICATIVE) {
-      if (rate.getFee() == null) {
-        Log.error("'fee' is missing in the GET /rate response");
-        throw new ServerErrorException("internal server error");
+    if (request.getType() == GetRateRequest.Type.FIRM) {
+      if (Objects.requireNonNull(rate).getFee() == null) {
+        logErrorAndThrow("'fee' is missing in the GET /rate response", ServerErrorException.class);
+      }
+      if (rate.getId() == null || rate.getExpiresAt() == null) {
+        logErrorAndThrow(
+            "'id' and/or 'expires_at' are missing in the GET /rate response",
+            ServerErrorException.class);
       }
     }
 
-    if (request.getType() == GetRateRequest.Type.FIRM) {
-      if (rate.getId() == null || rate.getExpiresAt() == null) {
-        Log.error("'id' and/or 'expires_at' are missing in the GET /rate response");
-        throw new ServerErrorException("internal server error");
+    // sell_amount is present and positive number
+    if (!isPositiveNumber(Objects.requireNonNull(rate).getSellAmount())) {
+      logErrorAndThrow(
+          "'sell_amount' is missing or not a positive number in the GET /rate response",
+          ServerErrorException.class);
+    }
+
+    // buy_amount is present and positive number
+    if (!isPositiveNumber(rate.getBuyAmount())) {
+      logErrorAndThrow(
+          "'buy_amount' is missing or not a positive number in the GET /rate response",
+          ServerErrorException.class);
+    }
+
+    // sell_amount has a proper number of significant decimals
+    if (!NumberHelper.hasProperSignificantDecimals(
+        Objects.requireNonNull(rate).getSellAmount(), sellAsset.getSignificantDecimals())) {
+      logErrorAndThrow(
+          "'sell_amount' has incorrect number of significant decimals in the GET /rate response",
+          ServerErrorException.class);
+    }
+
+    // buy_amount has a proper number of significant decimals
+    if (!NumberHelper.hasProperSignificantDecimals(
+        Objects.requireNonNull(rate).getBuyAmount(), buyAsset.getSignificantDecimals())) {
+      logErrorAndThrow(
+          "'buy_amount' has incorrect number of significant decimals in the GET /rate response",
+          ServerErrorException.class);
+    }
+
+    FeeDetails fee = rate.getFee();
+    // if fee is set, check the following
+    if (fee != null) {
+      // fee.total is present and is a positive number
+      if (!isPositiveNumber(fee.getTotal())) {
+        logErrorAndThrow(
+            "'fee.total' is missing or not a positive number in the GET /rate response",
+            ServerErrorException.class);
+      }
+      // fee.asset is a valid asset
+      AssetInfo feeAsset = assetService.getAssetByName(fee.getAsset());
```

Comment:
Should also check for fee's proper significant decimals

Reply:
Added.